addons/hr_holidays/models/hr_leave_type.py

```diff
@@ -52,10 +53,57 @@ class HolidaysType(models.Model):
     virtual_remaining_leaves = fields.Float(compute='_compute_leaves', string='Virtual Remaining Leaves',
         help='Maximum Leaves Allowed - Leaves Already Taken - Leaves Waiting Approval')
 
-    double_validation = fields.Boolean(string='Apply Double Validation',
+    double_validation = fields.Boolean(string='Apply Double Validation', compute='_compute_double_validation',
         help="When selected, the Allocation/Leave Requests for this type require a second validation to be approved.")
     company_id = fields.Many2one('res.company', string='Company', default=lambda self: self.env.user.company_id)
 
+    validation_type = fields.Selection([('hr', 'Human Resource Responsible'),
+                                      ('manager', 'Department Manager'),
+                                      ('both', 'Double Validation')],
+                                     default='hr',
+                                     string='Validation By')
+
+    sequence = fields.Integer(default=100,
+                              help='The type with the smallest sequence is the default value in leave request')
+
+    employee_visibility = fields.Selection([('both', 'On Leave As Well As On Allocation'),
+                                            ('leave', 'Only On Leave'),
+                                            ('allocation', 'Only On Allocation')],
+                                           default='both', string='Available For Employee :',
+                                           help='This leave type will be available on Leave / Allocation request based on selected value')
+
+    validity_start = fields.Date("Start Date", default=fields.Date.today(),
+                                 help='Adding validity to types of leaves so that it cannot be selected outside'
+                                 'this time period')
+    validity_stop = fields.Date("End Date")
+
+    valid = fields.Boolean(compute='_compute_valid', search='_search_valid', help='This indicates if it is still possible to use this type of leave')
+
+    @api.multi
+    @api.depends('validation_type')
+    def _compute_double_validation(self):
+        for holiday in self:
+            holiday.double_validation = holiday.validation_type == 'both'
+
+    @api.multi
+    @api.depends('validity_start', 'validity_stop', 'limit')
+    def _compute_valid(self):
```

Comment:
I don't fully understand the "valid" meaning. I am ok with the date management, it allows to avoid having to have complex domains in views, field domains, ...

However the limit part is strange. It means an holiday type without limit is not considered as "valid" and therefore will not be proposed to people. Taking our HR use as example, sick days or home working will be invalid if I understand well, unless you define some limit date in which case those are valid independently from the limit, which is strange.

In my opinion if we want to really take into account the holiday type limitation, we have the following cases (concerning limit only)
1/ no limit -> ok
2/ limit -> have to check already-encoded leave to see if the limit is bypassed
It could be based upon the remaining_leaves field.

Currently in the code you have some issues that prevent from really testing the limit part (default_get on leave / alloc that does not respect the field domain). We already have remaining leaves but as it is not stored, it is not searchable currently.

Anyway there is some functional cleaning to do, do we want to hide leave types in which we don't have any days left, or just display it with 0/20 remaining ? My opinion is that having void leave type does not mean we should not display it. Indeed I am glad to know I have already taken my 20 leave days. However displaying it in the bottom of the list could help people choosing the right leave type by default. It was done in the removed code about reordering by remaining leaves I think.